### PR TITLE
Feature/418 As a guardian i would like to be able to search for a specific weekplan so it is easier for me to find

### LINKED ISSUE #418: As a guardian i would like to be able to search for a specific weekplan so it is easier for me to find
**Describe the solution you'd like**
It should be possible to search for week plans by name, week, and/or year. Any week plan that does not match the search criteria should disappear, while still maintaining its normal sorting. 

**Additional context**
The solution should look like this:

**A search button should be added to the top bar.**
2020e_10_GUA_1_WP418 Expanded search bar

**After pressing the search button, a search bar should appear:**
2020e_10_GUA_2_WP418 Expanded search bar

**The user should then be able to type in the name, week, and/or year and get the resulting week plans:**
2020e_10_GUA_3_WP418 Expanded search bar

**The user can also choose to filter their search. As an example, removing the week number criteria reveals a week plan that is not in week 4:**
2020e_10_GUA_4_WP418 Expanded search bar

### PR DESCRIPTION
**Pull request by ggAlexS**

### CODE REVIEW

**Comment by Determinatoro** [comment]
Please use the following checklist, @LJensen0403. Feel free to check it off as you go along.

**Code Design**
- [x] The code is in the right place? (Both in terms of folder structure and class structure)
- [x] The code is not over-engineered. Examples of over-engineering: Implemented behavior for future problems
- [x] This code could not, to the best of my knowledge, have reused existing code
- [x] The code does not introduce functionality that is necessary for solving the problem

**Code Readability and Maintainability**
- [x] Names are meaningful and self-documenting
- [x] It is understandable, by reading the code, what it does
- [x] The code is simple and readable. i.e. it contains no "hack", or obscure solution
- [x] All new functionality is covered by tests
- [x] It is understandable, by reading the tests, what they do
- [x] The tests covers sensible cases. Both happy paths and exception paths
- [x] The tests cover the full functionality. i.e., The tests fails if some of the requested functionality is missing
- [x] Error messages in the code are understandable
- [x] Confusing sections of code are covered by comments

**Functionality**

For this part, open the app, and test:

- [x] The features that the code claims to implement, are actually implemented
- [x] The code does not look like it has bugs
- [x] All the new screens are reachable through in-app navigation

**Points for Discussion and Reflection**

- [ ] UI messages, that the user might meet, are actually helpful
- [ ] There are no errors that are bad practice. E.g. hard-coding a test database connection
- [x] There is well written code in this. Remember to tell the developer!

**Comment by Determinatoro** [comment]
Please use the following checklist, @ProgDaniel. Feel free to check it off as you go along.

**Code Design**
- [x] The code is in the right place? (Both in terms of folder structure and class structure)
- [x] The code is not over-engineered. Examples of over-engineering: Implemented behavior for future problems
- [x] This code could not, to the best of my knowledge, have reused existing code
- [x] The code does not introduce functionality that is necessary for solving the problem

**Code Readability and Maintainability**
- [x] Names are meaningful and self-documenting
- [x] It is understandable, by reading the code, what it does
- [x] The code is simple and readable. i.e. it contains no "hack", or obscure solution
- [x] All new functionality is covered by tests
- [x] It is understandable, by reading the tests, what they do
- [x] The tests covers sensible cases. Both happy paths and exception paths
- [x] The tests cover the full functionality. i.e., The tests fails if some of the requested functionality is missing
- [x] Error messages in the code are understandable
- [x] Confusing sections of code are covered by comments

**Functionality**

For this part, open the app, and test:

- [x] The features that the code claims to implement, are actually implemented
- [x] The code does not look like it has bugs
- [x] All the new screens are reachable through in-app navigation

**Points for Discussion and Reflection**

- [x] UI messages, that the user might meet, are actually helpful
- [x] There are no errors that are bad practice. E.g. hard-coding a test database connection
- [x] There is well written code in this. Remember to tell the developer!

**Review by LJensen0403 — CHANGES_REQUESTED** [review]
Hard-coded test database connection

**Comment by JeppeKLau** [comment]
> 
> 
> Hard-coded test database connection

Do elaborate

**Comment by LJensen0403** [comment]
Correct me if I'm wrong, but the line "SERVER_HOST": "http://10.0.2.2:5000" located in environment, means it connects to a database running locally. If I'm wrong, or if this is not an issue, I'd be more than happy to accept the pull request, as everything else looks good.

**Comment by ProgDaniel** [comment]
The "pubspec.lock", "pubspec.yaml", and "environment.json" should not be changed if it is not necessary. You should remember to clear changes in those files before you commit.

**Comment by JeppeKLau** [comment]
> 
> 
> The "pubspec.lock", "pubspec.yaml", and "environment.json" should not be changed if it is not necessary. You should remember to clear changes in those files before you commit.

I looked at the changes in `pubspec.yaml` and they are in order now. The added dependency `flappy_search_bar` is the implemented search bar used for the searching. 
`environment.json` should be reset. 
`pubspec.lock` automatically updates when running `flutter pub get` as this file tells which versions of the packages it was able to get a hold of.

**Review by LJensen0403 — DISMISSED** [review]

**Comment by ProgDaniel** [comment]
If a user has no "Overstående uger", that is Overstående uger has no items in it, then the searching won't progress, and it wont find any weekplans, even if there are any in the current week. (The loading symbol will remain swirling)

The created test, "search for weekplan", also does not pass.

File: lib/widgets/giraf_app_bar_widget.dart [hunk 1]
```diff
@@ -5,45 +5,58 @@ import 'package:weekplanner/models/enums/app_bar_icons_enum.dart';
...
-  GirafAppBar({Key key, this.title, this.appBarIcons, this.isGuardian = true})
+  GirafAppBar(
+      {Key key,
+      this.title,
+      this.appBarIcons,
+      this.isGuardian = true,
+      this.bottom})
       : toolbarBloc = di.getDependency<ToolbarBloc>(),
-        preferredSize = const Size.fromHeight(56.0),
+        preferredSize = const Size.fromHeight(53.0),
```
Comment (ProgDaniel) [inline lib/widgets/giraf_app_bar_widget.dart]: Is there a reason for this top-bar size change?

**Review by ProgDaniel — CHANGES_REQUESTED** [review]
Remember to discard "environments.json" before you commit, again 

Remove outcommented blocks of code, like line 41/42 in weekplan_selector_bloc.dart

**Comment by ProgDaniel** [comment]
The same bug as before still exists: if either new weeks or old weeks (overstående uger) does not have any weeks in it, you cannot search
image

The refactoring of the Observable was part of another issue, so you shouldnt change it in this issue, unless its necessary for the code to work.

**Comment by JeppeKLau** [comment]
> 
> 
> The same bug as before still exists: if either new weeks or old weeks (overstående uger) does not have any weeks in it, you cannot search
> image
> 
> The refactoring of the Observable was part of another issue, so you shouldnt change it in this issue, unless its necessary for the code to work.

Then you need to make sure you have recompiled the app after pulling the latest changes, because it works on my end when using a user with either no plans at all, only current plans, or both current and ended plans.

I only tried changing the Observable class to try and make the streams work better. It was simply part of the trial and error phase of getting this feature to work. If you would like I can change it back to Observable, but I don't really see a reason to given that it will be refactored by someone else.

**Review by ProgDaniel — DISMISSED** [review]

**Review by LJensen0403 — DISMISSED** [review]

**Review by LJensen0403 — DISMISSED** [review]

**Review by ProgDaniel — DISMISSED** [review]

**Review by ProgDaniel — APPROVED** [review]

**Review by LJensen0403 — APPROVED** [review]

**Pull request merged**